File: go/tests/unit/foundation_adaptive_bulkhead_test.go
package unit_test

import (
	"context"
	"testing"
	"time"

	"github.com/stretchr/testify/require"

	"github.com/gt-tech-ai/knowledge-engine/go/foundation/resilience/bulkhead"
	"github.com/gt-tech-ai/knowledge-engine/go/foundation/resilience/bulkhead/adaptive"
)

// TestAdaptiveBulkhead_LimitShrinksUnderLatencyAndRecovers tests that the AIMD limit shrinks
// under sustained latency and grows back once latency subsides.
//
// Why this test is important:
//   - The whole point of an adaptive limiter is to back off concurrency at a slowing backend
//     and reclaim it on recovery; if the limit did not move with latency it would be a fixed
//     bulkhead with extra machinery.
//
// What it tests:
//   - Four slow samples (RTT over the threshold) drive the limit below its initial value; five
//     fast samples then grow it back above the shrunk value.
func TestAdaptiveBulkhead_LimitShrinksUnderLatencyAndRecovers(t *testing.T) {
	t.Parallel()

	bh := adaptive.New(adaptive.Config{
		MinConcurrent:     1,
		MaxConcurrent:     20,
		InitialConcurrent: 10,
		RTTThreshold:      10 * time.Millisecond,
		BackoffRatio:      0.5,
	})
	require.Equal(t, 10, bh.Limit())

	for range 4 { // slow samples: RTT over the threshold shrinks the limit multiplicatively
		require.NoError(t, bh.Execute(context.Background(), func() error {
			time.Sleep(20 * time.Millisecond)
			return nil
		}))
	}
	shrunk := bh.Limit()
	require.Less(t, shrunk, 10, "sustained latency must shrink the adaptive limit")

	for range 5 { // fast samples: RTT under the threshold grows the limit back additively
		require.NoError(t, bh.Execute(context.Background(), func() error { return nil }))
	}
	require.Greater(t, bh.Limit(), shrunk, "the limit recovers once latency subsides")
}

// TestAdaptiveBulkhead_TryExecuteRejectsWhenFull tests that TryExecute rejects immediately once
// the current adaptive limit is saturated (the non-blocking contract of Bulkhead).
//
// Why this test is important:
//   - TryExecute exists so a caller can shed rather than queue; it must reject the moment the
//     limit is reached, or the "try" semantics are broken.
//
// What it tests:
//   - With a limit of 1 occupied by an in-flight Execute, a concurrent TryExecute returns
//     ErrBulkheadFull.
func TestAdaptiveBulkhead_TryExecuteRejectsWhenFull(t *testing.T) {
	t.Parallel()

	bh := adaptive.New(adaptive.Config{
		MinConcurrent:     1,
		MaxConcurrent:     1,
		InitialConcurrent: 1,
		RTTThreshold:      time.Hour, // never treat the held slot as a slow sample
		BackoffRatio:      0.9,
	})
	started := make(chan struct{})
	release := make(chan struct{})
	go func() {
		_ = bh.Execute(context.Background(), func() error {
			close(started)
			<-release
			return nil
		})
	}()

	<-started // the single slot is now occupied
	require.ErrorIs(
		t,
		bh.TryExecute(func() error { return nil }),
		adaptive.ErrBulkheadFull,
	)
	close(release)
}

// TestBulkhead_KindAdaptive_Selected tests that the factory selects the adaptive backend for
// KindAdaptive (the config-selects-impl contract, charter §2).
//
// Why this test is important:
//   - The adaptive limiter is opted into by a config Kind; if NewFromConfig silently returned a
//     channel bulkhead, the configured self-tuning behavior would never take effect.
//
// What it tests:
//   - bulkhead.New(KindAdaptive) returns an *adaptive.Bulkhead.
func TestBulkhead_KindAdaptive_Selected(t *testing.T) {
	t.Parallel()

	bh, err := bulkhead.New(bulkhead.KindAdaptive)
	require.NoError(t, err)
	_, ok := bh.(*adaptive.Bulkhead)
	require.True(t, ok, "KindAdaptive must select the adaptive backend")
}

// TestAdaptiveBulkhead_TryExecuteRunsWhenSlotAvailable tests that TryExecute runs the op and
// returns its result when a slot is free (the non-rejecting path), using the package defaults.
//
// Why this test is important:
//   - TryExecute must actually execute when capacity exists, not only reject when full; the happy
//     path is the common case and a regression there would silently drop work.
//
// What it tests:
//   - With a fresh default-config bulkhead, TryExecute invokes the op and returns nil.
func TestAdaptiveBulkhead_TryExecuteRunsWhenSlotAvailable(t *testing.T) {
	t.Parallel()

	bh := adaptive.New(adaptive.DefaultConfig())
	called := false
	require.NoError(t, bh.TryExecute(func() error {
		called = true
		return nil
	}))
	require.True(t, called, "TryExecute must run the op when a slot is available")
}

// TestAdaptiveBulkhead_ExecuteRespectsContextCancellation tests that a blocked Execute returns the
// context error when its context is cancelled while waiting for a slot (the ctx-aware acquire path).
//
// Why this test is important:
//   - A caller that cancels (deadline, shutdown) must not hang forever waiting for a saturated
//     bulkhead; honoring ctx during the wait is what makes the limiter safe under load.
//
// What it tests:
//   - With the single slot held, a second Execute blocks in acquire and returns context.Canceled
//     once its context is cancelled.
func TestAdaptiveBulkhead_ExecuteRespectsContextCancellation(t *testing.T) {
	t.Parallel()

	bh := adaptive.New(adaptive.Config{
		MinConcurrent:     1,
		MaxConcurrent:     1,
		InitialConcurrent: 1,
		RTTThreshold:      time.Hour,
		BackoffRatio:      0.9,
	})
	started := make(chan struct{})
	release := make(chan struct{})
	go func() {
		_ = bh.Execute(context.Background(), func() error {
			close(started)
			<-release
			return nil
		})
	}()
	<-started // the single slot is now held

	ctx, cancel := context.WithCancel(context.Background())
	done := make(chan error, 1)
	go func() { done <- bh.Execute(ctx, func() error { return nil }) }()
	cancel() // the blocked acquire must observe the cancellation and return

	require.ErrorIs(t, <-done, context.Canceled)
	close(release)
}

File: python/techai_webutils/tests/unit/foundation_adaptive_bulkhead_test.py
"""Tests for the adaptive (AIMD) bulkhead + its factory selection.

Parity with Go's ``go/tests/unit/foundation_adaptive_bulkhead_test.go``: the self-tuning limit
shrinks under sustained latency and recovers, TryExecute rejects when saturated, and the factory
selects the adaptive backend by kind.
"""

from __future__ import annotations

import asyncio

import pytest
from techai_webutils.foundation.resilience.bulkhead import (
    AdaptiveBulkhead,
    BulkheadConfig,
    BulkheadFullError,
    BulkheadKind,
    bulkhead_from_config,
)


@pytest.mark.asyncio
async def test_adaptive_limit_shrinks_under_latency_and_recovers() -> None:
    """The AIMD limit shrinks under sustained latency and grows back once latency subsides.

    Why this test is important:
        - The whole point of an adaptive limiter is to back off concurrency at a slowing backend
          and reclaim it on recovery; a limit that did not move with latency would be a fixed
          bulkhead with extra machinery.

    What it tests:
        - Four slow samples (RTT over the threshold) drive the limit below its initial value; five
          fast samples then grow it back above the shrunk value.
    """
    bh = AdaptiveBulkhead(
        min_concurrent=1,
        max_concurrent=20,
        initial_concurrent=10,
        rtt_threshold_seconds=0.01,
        backoff_ratio=0.5,
    )
    assert bh.limit == 10

    for _ in range(4):  # slow samples shrink the limit multiplicatively
        await bh.execute(lambda: asyncio.sleep(0.02))
    shrunk = bh.limit
    assert shrunk < 10, "sustained latency must shrink the adaptive limit"

    for _ in range(5):  # fast samples grow the limit back additively
        await bh.execute(lambda: asyncio.sleep(0))
    assert bh.limit > shrunk, "the limit recovers once latency subsides"


@pytest.mark.asyncio
async def test_adaptive_try_execute_rejects_when_full() -> None:
    """TryExecute rejects immediately once the current adaptive limit is saturated.

    Why this test is important:
        - TryExecute exists so a caller can shed rather than queue; it must reject the moment the
          limit is reached, or the "try" semantics are broken.

    What it tests:
        - With a limit of 1 occupied by an in-flight execute, a concurrent try_execute raises
          BulkheadFullError.
    """
    bh = AdaptiveBulkhead(
        min_concurrent=1,
        max_concurrent=1,
        initial_concurrent=1,
        rtt_threshold_seconds=3600,  # never treat the held slot as a slow sample
        backoff_ratio=0.9,
    )
    started = asyncio.Event()
    release = asyncio.Event()

    async def held() -> None:
        started.set()
        await release.wait()

    task = asyncio.create_task(bh.execute(held))
    await started.wait()  # the single slot is now occupied

    with pytest.raises(BulkheadFullError):
        await bh.try_execute(lambda: asyncio.sleep(0))

    release.set()
    await task


@pytest.mark.asyncio
async def test_bulkhead_from_config_selects_adaptive() -> None:
    """The factory selects the adaptive backend for ADAPTIVE (config-selects-impl contract).

    Why this test is important:
        - The adaptive limiter is opted into by a config kind; if the factory silently returned a
          semaphore bulkhead, the configured self-tuning behavior would never take effect.

    What it tests:
        - ``bulkhead_from_config`` with ``BulkheadKind.ADAPTIVE`` returns an ``AdaptiveBulkhead``.
    """
    bh = bulkhead_from_config(BulkheadConfig(kind=BulkheadKind.ADAPTIVE))
    assert isinstance(bh, AdaptiveBulkhead)


def test_bulkhead_from_config_unknown_kind_raises() -> None:
    """The factory raises on an unknown kind rather than silently returning a degenerate bulkhead.

    Why this test is important:
        - A factory that swallowed an unknown kind would ship a mis-selected policy silently;
          failing loudly surfaces the config error at startup.

    What it tests:
        - ``bulkhead_from_config`` raises ``ValueError`` for a kind outside the enum.
    """
    bad = BulkheadConfig.__new__(BulkheadConfig)
    object.__setattr__(bad, "kind", "bogus")  # the unknown-kind path reads only .kind

    with pytest.raises(ValueError, match="unknown bulkhead kind"):
        bulkhead_from_config(bad)
